packages/nimble-components/src/table/specs/table-column-sort-hld.md

```diff
@@ -0,0 +1,75 @@
+# Table Column Sorting HLD
+
+## Problem Statement
+
+There is a requirement to have table columns that can be sorted, both programmatically and interactively. This document focuses specifically on programmatic sorting of columns.
+
+### Out of scope of this HLD
+
+While some interactive sorting may be alluded to within this document, the details of interactive sorting are out of scope for this document. For example, this document does not cover:
+
+-   How a column will opt in/out of being interactively sortable
+-   Event(s) related to sorting
+-   Interactions that lead to the sorting state changing, e.g. clicking a header, option(s) in the header's menu
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#874 Programmatically sort columns](https://github.com/ni/nimble/issues/874)
+-   [#854 Interactively sort the columns in a table](https://github.com/ni/nimble/issues/854)
+-   [Table README](./README.md)
+-   [Table Design Doc](https://xd.adobe.com/view/5b476816-dad1-4671-b20a-efe796631c72-0e14/screen/d389dc1e-da4f-4a63-957b-f8b3cc9591b4/specs/)
+
+## Implementation / Design
+
+### API
+
+The sorting state of the table will be declaritively defined within the `TableColumn` elements that are slotted within the table. A column will optionally be able to define its sort direction (i.e. `ascending` vs `descending`) and its sort index. The sort index enables a table to be sorted by multiple columns at the same time. For example, if a table should be sorted by a first name column and then a last name column, the first name column would have a sort index of 0, and the last name column would have a sort index of 1. This is a similar API to AG Grid, which provides [sort](https://www.ag-grid.com/javascript-data-grid/column-state/#reference-ColumnStateParams-sort) and [sortIndex](https://www.ag-grid.com/javascript-data-grid/column-state/#reference-ColumnStateParams-sortIndex) configurations.
+
+In some cases, a client may want to configure their table to be sorted by a column that is not visible within the table. To enable this, a column within the table can be marked as hidden. A hidden column exists in the table to provide metadata that can be used for operations, such as sorting, but it is not rendered within the table.
```

Comment:
It's not in SLE at this point in time, but it's in Leslie's wireframes. The grid configuration slide out should have 3 tabs (currently there are only 2). Those tabs are (1) visible columns, (2) sorted columns, (3) grouping. Given that long-term vision, I think it makes sense to allow sorting by non-visible columns rather than requiring clients to implement some workaround involving modifying the `operationalDataFieldName`. Similarly, it will be a requirement to group by non-visible columns (already a requirement in SLE), so I expect the concept of hidden columns will also be leveraged for that.

Reply:
Gotcha, I agree that use case would require explicit hidden columns.